### PR TITLE
Do not remove containers from stats list on err

### PR DESCRIPTION
**Pull request by cpuguy83**
Before this patch, containers are silently removed from the stats list
on error. This patch instead will display `--` for all fields for the
container that had the error, allowing it to recover from errors.

When containers have an error, they will see:
image

Note: only the container with the error will have the `err` message, for the case of the screenshot I forced the error.

### CODE REVIEW

File: integration-cli/docker_cli_stats_test.go [hunk 1]
```diff
@@ -97,10 +97,7 @@ func (s *DockerSuite) TestStatsAllNoStream(c *check.C) {
 
 func (s *DockerSuite) TestStatsAllNewContainersAdded(c *check.C) {
 	// Windows does not support stats
-	// TODO: remove SameHostDaemon
```
Comment (cpuguy83) [inline integration-cli/docker_cli_stats_test.go]: Removed so I can run win2lin against this more. It may very well resolve this issue.

**Comment by duglin** [comment]
@cpuguy83 what defines an "error" ?  non-zero exit code doesn't seem to do it for me.

**Comment by cpuguy83** [comment]
@duglin it's an error while collecting stats for a container... So a decode error or an error from the call to the stats endpoint for a container.

**Comment by duglin** [comment]
Ah, gotcha.  Makes me wonder if we should treat the inability to connect to the daemon as one of these errors too.  Then a "docker stats" won't die when I recycle my daemon.

**Comment by duglin** [comment]
And if so, it might be better to just replace all lines in the table with something like:
```
CONTAINER           CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O
Error retrieving data, will keep trying...
```

**Comment by thaJeztah** [comment]
> And if so, it might be better to just replace all lines in the table with something like:

I think I like the original output as it's easier on the eye. Perhaps `--/--` instead of `err/err`? (Can we use colors?)

**Comment by calavera** [comment]
I think I like `--/--` better too, specially if it's something the command can recover of.

**Comment by thaJeztah** [comment]
this needs a rebase; also can you change this to the `--/--` format?

**Comment by cpuguy83** [comment]
Works for me, here's what it looks like:

image

**Comment by cpuguy83** [comment]
updated

**Comment by duglin** [comment]
@cpuguy83 I'm just curious, how did you first come across this issue?  Seems like it would be really hard to see in normal usage.

**Comment by cpuguy83** [comment]
@duglin Agreed, not likely to find this. I've not experienced it, but saw it while trying to figure out why the test for `docker stats` (no args) fails on win2lin (which we're just skipping right now).

**Comment by cpuguy83** [comment]
Also no, I don't think this fixes the win2lin issue.

**Comment by duglin** [comment]
@cpuguy83 interesting - thanks for the info

**Comment by thaJeztah** [comment]
+1 on design; @duglin are you ok with it like this? If you are, can you move it to "code review"?

**Comment by duglin** [comment]
SGTM

**Comment by duglin** [comment]
@cpuguy83 what's the best way for me to test this?  I guess to add some kind of failure logic, but where's the best spot to add it?

File: api/client/stats_helpers.go [hunk 2]
```diff
@@ -167,10 +169,17 @@ func (s *containerStats) Collect(cli client.APIClient, streamStats bool, waitFir
 func (s *containerStats) Display(w io.Writer) error {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
+	// NOTE: if you change this format, you must also change the err format below!
+	format := "%s\t%.2f%%\t%s / %s\t%.2f%%\t%s / %s\t%s / %s\t%d\n"
 	if s.err != nil {
```
Comment (cpuguy83) [inline api/client/stats_helpers.go]: @duglin to test, change this to `==`

**Comment by duglin** [comment]
@cpuguy83 I might be doing something wrong but as I was testing this I decided to force an error at https://github.com/docker/docker/pull/20835/files#diff-234033fe3b8730c35e53ce0d8dbda485R95 1/3 of the time. And I noticed that once there was an error decoding the stats for a container, while it did show "--/--" for the output, it never went back to showing the real stats. I think the code is assuming that once a container is either dead or has an error getting the stats it should stop checking it - but for the cases you're concerned about I think it should keep trying.  Any thoughts on this?

**Comment by duglin** [comment]
I added it at line 95 of stats_helper.go, on the `dec.Decode(&v)` call.

**Comment by cpuguy83** [comment]
@duglin You're right, there were 2 places this was an issue. Fixed.

**Comment by duglin** [comment]
@cpuguy83 I do now see the `--`s come and go randomly so that's cool.  However, the first time I ran it it didn't show new containers as they were added. But then suddenly it started to work the next time I ran "docker stats".  Still investigating....  but thought I'd share the info in case you have a guess as to what it might be.

**Comment by thaJeztah** [comment]
@cpuguy83 looks like CI failed;

```
14:33:19 # github.com/docker/docker/api/client
14:33:19 api/client/stats_unit_test.go:23: cannot use sync.RWMutex literal (type sync.RWMutex) as type sync.Mutex in field value
```

**Comment by duglin** [comment]
I've been able to reproduce the issue (container list always being empty) but but not reliably, it appears random. On the positive side it happens even w/o this PR so its not related to this code change. 
On to code review...

File: api/client/stats_helpers.go [hunk 3]
```diff
@@ -150,7 +154,7 @@ func (s *containerStats) Collect(cli client.APIClient, streamStats bool, waitFir
 				s.mu.Lock()
 				s.err = err
 				s.mu.Unlock()
-				return
+				continue
```
Comment (duglin) [inline api/client/stats_helpers.go]: Up above we'll show zeros if we don't get a response within 2 seconds, I'm wondering if we should show --'s now instead?

File: api/client/stats_helpers.go [hunk 4]
```diff
@@ -165,12 +169,21 @@ func (s *containerStats) Collect(cli client.APIClient, streamStats bool, waitFir
...
+	format := "%s\t%.2f%%\t%s / %s\t%.2f%%\t%s / %s\t%s / %s\t%d\n"
 	if s.err != nil {
-		return s.err
+		format = "%s\t%s\t%s / %s\t%s\t%s / %s\t%s / %s\t%s\n"
+		errStr := "--"
+		fmt.Fprintf(w, format,
+			s.Name, errStr, errStr, errStr, errStr, errStr, errStr, errStr, errStr, errStr,
+		)
+		err := s.err
+		s.err = nil
```
Comment (duglin) [inline api/client/stats_helpers.go]: Should we really be setting s.err to nil here? I would it think be better to leave it in the error state until we get a successful parse of data from the daemon.

**Comment by thaJeztah** [comment]
ping @cpuguy83 were you still working on this?

**Comment by cpuguy83** [comment]
Yep, just low priority for me at the moment.

**Comment by cpuguy83** [comment]
Ok, this is updated, and green.

**Comment by calavera** [comment]
LGTM

**Comment by thaJeztah** [comment]
ping @duglin PTAL

File: api/client/stats_helpers.go (hunk 3 above)
Comment (duglin) [inline api/client/stats_helpers.go]: @cpuguy83 what do you think about this?  I'm not sure why we're treating a "no response within 2 seconds" differently from an error case. In both cases we don't have data so showing --'s seems better than potentially misleading people by showing zeros.

Comment (cpuguy83) [inline api/client/stats_helpers.go]: I can change this behavior if we really want to, but is beyond the intention of the PR.

Comment (duglin) [inline api/client/stats_helpers.go]: we can do it in a different PR if you want, it just seemed like a good thing to do here since in both cases, from the end user perspective, we're dealing with the error case of not getting the data we expected back. up to you...

Comment (cpuguy83) [inline api/client/stats_helpers.go]: Fixed. Ended up being a one-liner change.

Comment (duglin) [inline api/client/stats_helpers.go]: Technically we don't need to set everything to zero any more, but that's just noise - thanks for the update!

**Comment by cpuguy83** [comment]
ping @duglin Do this LGTY?

**Comment by duglin** [comment]
Sorry - got distracted - will do a final test/review tomorrow am.

**Comment by duglin** [comment]
please test this

**Comment by duglin** [comment]
please test this, pretty please  :-)

**Comment by duglin** [comment]
assuming the doc issue isn't real
LGTM

**Comment by thaJeztah** [comment]
"retest this please" is the magic word, but it sometimes doesn't work

**Comment by thaJeztah** [comment]
Oh,and there's no doc changes here, so it's just Jenkins wrongly starting the docs CI
